File: php/src/LombokAlgoritma/Hash/Murmur3.php
<?php

// LombokAlgoritma — MurmurHash3_x86_32 (SPEC §12)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

declare(strict_types=1);

namespace LombokAlgoritma\Hash;

use LombokAlgoritma\Core\U32;

/** MurmurHash3_x86_32 (Austin Appleby, public-domain reference). Non-cryptographic. */
final class Murmur3
{
    private const C1 = 0xcc9e2d51;
    private const C2 = 0x1b873593;

    /** MurmurHash3_x86_32 of `$data` with a u32 seed, as an unsigned int. */
    public static function hash32(string $data, int $seed = 0): int
    {
        $h = $seed & U32::MASK;
        $n = strlen($data);
        $n4 = $n & ~3;
        for ($i = 0; $i < $n4; $i += 4) {
            $k = U32::mul(U32::rotl(U32::mul(U32::readLe($data, $i), self::C1), 15), self::C2);
            $h ^= $k;
            $h = U32::add(U32::mul(U32::rotl($h, 13), 5), 0xe6546b64);
        }
        $rem = $n - $n4;
        if ($rem > 0) {
            $k = 0;
            for ($i = 0; $i < $rem; $i++) {
                $k |= ord($data[$n4 + $i]) << (8 * $i);
            }
            $h ^= U32::mul(U32::rotl(U32::mul($k, self::C1), 15), self::C2);
        }
        return self::fmix32($h ^ ($n & U32::MASK));
    }

    /** MurmurHash3 32-bit finaliser (fmix32). */
    public static function fmix32(int $h): int
    {
        $h &= U32::MASK;
        $h ^= $h >> 16;
        $h = U32::mul($h, 0x85ebca6b);
        $h ^= $h >> 13;
        $h = U32::mul($h, 0xc2b2ae35);
        return $h ^ ($h >> 16);
    }
}

File: rust/lombokalgoritma/src/string/hash/murmur3.rs
// LombokAlgoritma — MurmurHash3_x86_32 (SPEC §12)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

/// `MurmurHash3_x86_32` (Austin Appleby) with a 32-bit seed.
pub fn murmur3_32(data: &[u8], seed: u32) -> u32 {
    const C1: u32 = 0xcc9e_2d51;
    const C2: u32 = 0x1b87_3593;
    let mut h = seed;
    let mut chunks = data.chunks_exact(4);
    for c in &mut chunks {
        let mut k = u32::from_le_bytes([c[0], c[1], c[2], c[3]]);
        k = k.wrapping_mul(C1).rotate_left(15).wrapping_mul(C2);
        h ^= k;
        h = h.rotate_left(13).wrapping_mul(5).wrapping_add(0xe654_6b64);
    }
    let tail = chunks.remainder();
    if !tail.is_empty() {
        let mut k = 0u32;
        for (i, &b) in tail.iter().enumerate() {
            k |= u32::from(b) << (8 * i);
        }
        h ^= k.wrapping_mul(C1).rotate_left(15).wrapping_mul(C2);
    }
    h ^= data.len() as u32;
    fmix32(h)
}

/// `MurmurHash3` 32-bit finalizer (`fmix32`).
pub fn fmix32(mut h: u32) -> u32 {
    h ^= h >> 16;
    h = h.wrapping_mul(0x85eb_ca6b);
    h ^= h >> 13;
    h = h.wrapping_mul(0xc2b2_ae35);
    h ^ (h >> 16)
}
